### PR TITLE
Maintain scrollbar position during a resize operation

### LINKED ISSUE #3494: Maintain scrollbar position during resize
# Description of the new feature/enhancement

Currently, if you drag back the scrollbar and then resize the window, the viewport jumps back to the "regular" position (the end of the data buffer, and the bottom position of the scrollbar).

It would be better to (approximately) maintain the scrolled position, so that the user keeps seeing whichever data was previously shown.

In VTE we usually track the position at the bottom of the viewport and try to keep that at the bottom; except when fully scrolled back to the top. More details in [doc/rewrap.txt](https://gitlab.gnome.org/GNOME/vte/blob/master/doc/rewrap.txt).

### PR DESCRIPTION
**Pull request by zadjii-msft**
## Summary of the Pull Request

Currently, when the user resizes the Terminal, we'll snap the visible viewport back to the bottom of the buffer. This PR changes the visible viewport of the Terminal to instead remain in the same relative location it was before the resize.  

## References
Made possible by our sponsors at #4741, and listeners like you. 

## PR Checklist
* [x] Closes #3494
* [x] I work here
* [ ] Tests added/passed
* [n/a] Requires documentation to be updated

## Detailed Description of the Pull Request / Additional comments

We already hated the `std::optional<short>&` thing I yeet'd into #4741 right at the end to replace a `short*`. So I was already going to change that to a `std::optional<std::reference_wrapper<short>>`, which is more idomatic. But then I was looking through the list of bugs and #3494 caught my eye. I realized it would be trivial to not only track the top of the `mutableViewport` during a resize, but we could use the same code path to track the _visible_ viewport's start as well. 

So basically I'm re-using that bit of code in `Reflow` to calculate the visible viewport's position too.

## Validation Steps Performed

Gotta love just resizing things all day, errday

### CODE REVIEW

File: src/buffer/out/textBuffer.cpp [hunk 1]
```diff
@@ -1869,15 +1869,15 @@ std::string TextBuffer::GenRTF(const TextAndColor& rows, const int fontHeightPoi
 // - lastCharacterViewport - Optional. If the caller knows that the last
 //   nonspace character is in a particular Viewport, the caller can provide this
 //   parameter as an optimization, as opposed to searching the entire buffer.
-// - oldViewportTop - Optional. The caller can provide a row in this parameter
-//   and we'll calculate the position of the _end_ of that row in the new
-//   buffer. The row's new value is placed back into this parameter.
+// - oldRows - Optional. The caller can provide a pair of rows in this parameter
```
Comment (DHowett-MSFT) [inline src/buffer/out/textBuffer.cpp]: nit: i'd call it something like "positionInformation" or name the struct that -- it's about resizing position anyway yeah?

Comment (zadjii-msft) [reply src/buffer/out/textBuffer.cpp]: Yea it's not a great name, I agree

**Comment by DHowett-MSFT** [comment]
We won't be able to clear the one failed build, so I'm going to admin merge this when it's time.

File: src/buffer/out/textBuffer.hpp [hunk 2]
```diff
@@ -161,10 +161,16 @@ class TextBuffer final
...
+    {
+        short mutableViewportTop{ 0 };
+        short visibleViewportTop{ 0 };
+    };
+
     static HRESULT Reflow(TextBuffer& oldBuffer,
                           TextBuffer& newBuffer,
                           const std::optional<Microsoft::Console::Types::Viewport> lastCharacterViewport,
-                          std::optional<short>& oldViewportTop);
+                          std::optional<std::reference_wrapper<PositionInformation>> positionInfo);
```
Comment (carlos-zamora) [inline src/buffer/out/textBuffer.hpp]:
```suggestion
                          std::optional<std::reference_wrapper<PositionInformation>> positionInfo = std::nullopt);
```
I think you can do it for the `lastCharacterViewport` parameter too.

Then the call you have in screenInfo doesn't have two `std::nullopt`s passed through.

Up to you tho

File: src/cascadia/TerminalCore/Terminal.cpp [hunk 3]
```diff
@@ -196,12 +201,28 @@ void Terminal::UpdateSettings(winrt::Microsoft::Terminal::Settings::ICoreSetting
...
 
-        std::optional<short> oldViewStart{ oldViewportTop };
+        // Build a PositionInformation to track the position of both the top of
+        // the mutable viewport and the top of the visible viewport in the new
+        // buffer.
+        // * the new value of mutableViewportTop will be used to figure out
+        //   where we should place the mutable viewport in the new buffer. This
+        //   requires a bit of trickiness to remain consistent with conpty's
+        //   buffer (as seen below).
+        // * then new value of visibleViewportTop will be used to calculate the
```
Comment (carlos-zamora) [inline src/cascadia/TerminalCore/Terminal.cpp]:
```suggestion
        // * the new value of visibleViewportTop will be used to calculate the
```

File: src/cascadia/TerminalCore/Terminal.cpp [hunk 4]
```diff
@@ -196,12 +201,28 @@ void Terminal::UpdateSettings(winrt::Microsoft::Terminal::Settings::ICoreSetting
...
+        //   new scrollOffsett in the new buffer, so that the visible lines on
+        //   the sceren remain roughly the same.
```
Comment (carlos-zamora) [inline src/cascadia/TerminalCore/Terminal.cpp]:
```suggestion
        //   the screen remain roughly the same.
```
